how-does-fix-with-copilot-work.md

```diff
@@ -0,0 +1,136 @@
+<!-- see inlineChatCommand.ts:registerInlineChatCommands
+and inlineChatCodeActions.ts:QuickFixesProvider -->
+
+When you click on *Fix With Copilot*, it invokes Copilot's inline chat with a hidden prompt constructed from JSX components. JSX&mdash;but not React. Copilot has its own renderer that generates plain text from JSX components, so *Fix With Copilot* is structured as a React-like hierarchy of small components. In addition to looking familiar, one clever property on all these components is `priority`. The renderer can drop low-priority components if the prompt gets too big. Here's an example of what the code looks like. `diagnostics` is the list of errors to be fixed:
+<!-- see inlineChatFix4Prompt.tsx (old version for now) -->
+
+```tsx
+diagnostics.map((d, idx) => {
+    const cookbook = getFixCookbook(documentContext.language.languageId, d);
+    return <>
+        <DiagnosticDescription 
+          diagnostic={d} 
+          cookbook={cookbook} 
+          showLineNumbers={this.props.showLineNumbers ?? false} 
+          maxLength={LINE_CONTEXT_MAX_SIZE} 
+          documentContext={documentContext} />
+        <DiagnosticRelatedInfo diagnostic={d} cookbook={cookbook} document={documentContext.document} />
+        <DiagnosticSuggestedFix cookbook={cookbook} />
+    </>;
+})
+```
+
+What components are there? The main ones are:
+
+- Include code surrounding the error.
+- Format the error nicely.
+- Provide instructions to Copilot.
+
+By default *Fix With Copilot* includes the entire file around the error. Of course, most of this time the entire file won't fit in the prompt. Instead, it includes the code in current function and summarizes code outside. The summary is constructed by X.
+
+So next it tries to summarise the code outside the current function by selectively deleting code. By default it deletes all function bodies outside the current one.
+(By default it divides the file into blocks and builds a tree out of it. Then each block gets assigned a weight based on distance from the error, and nesting depth. Blocks below a cutoff are omitted, so the prompt will include code near the error and less and less detail further away. 
+
+    TODO: Find out from Martin how commonly this method is the one used, probably only describe the simple or complex one. This would be a good place for a diagram to explain the complex method.
+
+Formatting the error starts with the error text itself, unless my cookbook offers a replacement. But it also includes a couple of sub-components. The first includes code from related error spans. Language services can attach related information to an error, and Copilot will include the function that contains each related location. The other subcomponent is example usages, which is what I showed earlier for fixing Typescript errors. TODO: Make this longer, better structured and more readable.
```

Comment:
The cookbook is implicitly mentioned in the code above, but this is the first mention otherwise.

Reply:
```suggestion
Formatting the error starts with the error text itself (unless my cookbook offers a replacement). But it also includes a couple of sub-components. The first includes code from related error spans. Language services can attach related information to an error, and Copilot will include the function that contains each related location. The other subcomponent is example usages, which is what I showed earlier for fixing Typescript errors. TODO: Make this longer, better structured and more readable.
```

Reply:
Yeah, I decided to go with 'nudge' and 'instructions' rather than cookbook. I'm not sure it's the right choice but cookbook doesn't seem quite accurate to me either.